serialization/object_serialization.nim

```diff
@@ -224,9 +224,15 @@ proc makeFieldReadersTable(RecordType, Reader: distinct type):
 proc fieldReadersTable*(RecordType, Reader: distinct type):
                         ptr seq[FieldReader[RecordType, Reader]] =
   mixin readValue
-  var tbl {.global.} = makeFieldReadersTable(RecordType, Reader)
-  {.gcsafe.}:
-    return addr(tbl)
+
+  # careful: https://github.com/nim-lang/Nim/issues/17085
+  # TODO why is this even here? one could just return the function pointer
```

Comment:
`.global.` variable are often used to generate one constant or other RTTI object *per type*. Normal constants and global variables don't have the property of being "instantiatable" like a generic. So, the function creates one unique global for each type and returns its address. The C compiler will notice the constant return value and it will erase the function replacing all calls to the address of the generated global.

The actual code that uses the table to locate the reader for a particular field at run-time sits elsewhere.

Reply:
```
proc findFieldReader*(RecordType, Reader: distinct type,
                      wantedName: string,
                      expectedFieldPos: var int): auto =
  var x = 0
  enumAllSerializedFields(RecordType):
    proc readField(obj: var RecordType, reader: var Reader)
                  {.gcsafe, nimcall, raises: [SerializationError, Defect], locks: 0.} =
      when RecordType is tuple:
        const i = fieldName.parseInt
      try:
        type F = FieldTag[RecordType, realFieldName, type(FieldType)]
        when RecordType is tuple:
          obj[i] = readFieldIMPL(F, reader)
        else:
          # TODO: The `FieldType` coercion below is required to deal
          # with a nim bug caused by the distinct `ssz.List` type.
          # It seems to break the generics cache mechanism, which
          # leads to an incorrect return type being reported from
          # the `readFieldIMPL` function.
          field(obj, realFieldName) = FieldType readFieldIMPL(F, reader)
      except SerializationError as err:
        raise err
      except CatchableError as err:
        reader.handleReadException(
          `RecordType`,
          fieldName,
          when RecordType is tuple: obj[i] else: field(obj, realFieldName),
          err)

    if x > expectedFieldPos: 
      if fieldName == wantedName:
        return readField
   # repeat for the other fields
  return nil
```

this also finds the reader, without a global - ie the global, to the best of my understanding, is used to work around the poor `enumAllSerializedFields` api?

Reply:
Well, this is not really doing the same, although it's close enough to probably not matter in practice.
I'll merge this simple PR now, but will revisit the problem once 1.2.10 is out.